api/v1alpha1/database_webhook.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2022-2023 Nutanix, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"regexp"
+
+	"github.com/nutanix-cloud-native/ndb-operator/api"
+	"github.com/nutanix-cloud-native/ndb-operator/common"
+	"github.com/nutanix-cloud-native/ndb-operator/common/util"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var databaselog = logf.Log.WithName("database-resource")
+
+func (r *Database) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-ndb-nutanix-com-v1alpha1-database,mutating=true,failurePolicy=fail,sideEffects=None,groups=ndb.nutanix.com,resources=databases,verbs=create;update,versions=v1alpha1,name=mdatabase.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &Database{}
+
+func instanceSpecDefaulterForCreate(r *Database) {
+
+	if len(r.Spec.Instance.DatabaseNames) == 0 {
+		r.Spec.Instance.DatabaseNames = api.DefaultDatabaseNames
+	}
+
+	if r.Spec.Instance.TimeZone == "" {
+		r.Spec.Instance.TimeZone = "UTC"
+	}
+
+	if r.Spec.Instance.Profiles.Compute.Id == "" && r.Spec.Instance.Profiles.Compute.Name == "" {
+		r.Spec.Instance.Profiles.Compute.Name = common.PROFILE_DEFAULT_OOB_SMALL_COMPUTE
+	}
+
+	// time machine defaulter logic
+
+	if r.Spec.Instance.TMInfo.Name == "" {
+		r.Spec.Instance.TMInfo.Name = r.Spec.Instance.DatabaseInstanceName + "_TM"
+	}
+
+	if r.Spec.Instance.TMInfo.Description == "" {
+		r.Spec.Instance.TMInfo.Description = "Time Machine for " + r.Spec.Instance.DatabaseInstanceName
```

Comment:
How can we also handle the case when the databaseInstanceName is empty?

Reply:
We are not defaulting to any name for DatabaseInstanceName so the webhook flow will take care of it. The validation webhook will fail down the line if DatabaseInstanceName is empty.